Add guidance on including api-version in nextLink (#376)
Microsoft/api-guidelines: azure/ConsiderationsForServiceDesign.md

Finally, write sample code for your service's workflow and add the code you'd want customers using to gracefully recover from errors.  Is it actually graceful?  Is it something you'd be comfortable asking most customers to write?  We also highly encourage reaching out to customers during private preview and asking them for code they've written against your service.  Their error handling might match your expectations, you might find a strong need for better documentation, or you might find important opportunities to improve the errors you're returning.

## Pagination

> Operations that return a collection of resources may need should support pagination.

Reviewer: ```suggestion
Operations that return a collection of resources must consider pagination.
```

There are hard limits to the payload size of HTTP responses, and when the size of a collection or the resources themselves
can grow arbitrarily large there is the risk of exceeding this limit if the operation does not support pagination.
Further, adding support for pagination is a breaking change so it should be supported in the initial GA of the service
if there is any possibility that it will eventually be needed.

There are two forms of pagination that MAY be supported by RESTful APIs.
Server-driven paging mitigates against denial-of-service attacks by forcibly paginating a request over multiple response payloads.
Client-driven paging enables clients to request only the number of resources that it can use at a given time.

> 

Reviewer: Services almost always want to support server-driven paging and can optionally support client-driven paging as well.

> 

Author: ```suggestion
Services should almost always support server-driven paging and may optionally support client-driven paging.

```

### Server-driven paging

In server-driven paging, the service includes a `nextLink` property in the response to indicate that additional elements
exist in the collection.

> The value of the `nextLink` property should be an opaque absolute URL that will return the next page of results.

Reviewer: Can we say "practically opaque" or "almost opaque" or maybe "semi-opaque" URL because it is not completely opaque.

The absence of a `nextLink` property means that no additional pages are available.

> Since `nextLink` is an opaque URL it should include any query parameters required by the service, including `api-version`.

Reviewer: Do you not want to say anything about clients replacing the api-version query parameter? I think we should so that service teams know what the expected client usage is.

Author: We probably should say something about the service behavior when a client changes _something_ in the nextLink URL. We could say:
> The service should honor a request to a URL derived from `nextLink` by replacing the value for the `apl-version` query parameter with a different but valid api version.  The service may reject the request if any other element of `nextLink` was modified.

What do you think of that?

Reviewer: The service MAY reject the nextLink request if any element of nextLink was modified by the client except for the api-version query parameter as the client should change this to reflect the version that it expects. Of course, the service must reject the request if the client passes an api-version value not supported by the service.

Author: I said "should" rather than "must" because I think there may be cases where changing the api-version creates an invalid request.  For example, suppose a service adds support for "filter" on a pageable GET operation in V2. A V2 client makes a request to this endpoint with `filter=something`. The `nextLink` in the response has `?filter=something&continuationToken=xxxyyy&api-version=v2`.  Then the client modifies the URL to have `api-version=v1` -- which does not support `filter`. How could we allow the service to return a successful response for this?

Reviewer: We're designing our SDKs to always replace the api-version. How would this work with our SDKs? If the filter were embedded in the continuationToken, then the service could make it work transparently to the client. Alternatively, the V1-service could support the V2-filter in a non-breaking way. The end-result is that the service HAS TO make this work transparently to the client via some opaque thing (like a continuationToken or the opaque part of the URL/query params).

This is more reason to prefer continuationTokens over nextLink. Weren't you going to write a proposal for this?
